cli/transmissioncli.1
.\"
.\"  Copyright (c) Deanna Phillips <deanna@sdf.lonestar.org>
.\" 
.\"  Permission to use, copy, modify, and distribute this software for any
.\"  purpose with or without fee is hereby granted, provided that the above
.\"  copyright notice and this permission notice appear in all copies.
.\" 
.\"  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\"  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\"  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\"  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\"  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\"  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\"  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd July 21, 2008
.Dt TRANSMISSIONCLI 1
.Os
.Sh NAME
.Nm transmissioncli
.Nd a bittorrent client
.Sh SYNOPSIS
.Nm transmissioncli
.Bk -words
.Fl h
.Nm
.Fl i
.Ar torrent-file
.Nm
.Fl s
.Ar torrent-file
.Nm
.Fl n Ar sourcefile
.Op Fl a Ar url
.Op Fl c Ar comment
.Op Fl r
.Ar new-torrent-file
.Nm
.Op Fl b | B
.Op Fl d Ar number | Fl D
.Op Fl er | ep | et
.Op Fl f Ar script
.Op Fl g Ar directory
.Op Fl h
.Op Fl m | M
.Op Fl p Ar port
.Op Fl t
.Op Fl u Ar number | Fl U
.Op Fl v
.Op Fl w Ar directory
.Ar torrent-file
.Ek
.Sh DESCRIPTION
The
.Nm
program is a lightweight, command-line BitTorrent client with
scripting capabilities.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl a, Fl -announce Ar announce-url
When creating a new torrent, set its announce URL
.It Fl b Fl -blocklist
Enable peer blocklists.  Transmission understands the bluetack blocklist file format.
New blocklists can be added by copying them into the config-dir's "blocklists" subdirectory.
.It Fl B Fl -no-blocklist
Disble blocklists.
.It Fl c, Fl -comment Ar comment-text
When creating a new torrent, set its comment field
.It Fl d, -downlimit Ar number
Set the maximum download speed in KB/s
.It Fl D, -no-downlimit
Don't limit the download speed
.It Fl er Fl -encryption-required
Encrypt all peer connections.
.It Fl ep Fl -encryption-preferred 
Prefer encrypted peer connections.
.It Fl et Fl -encryption-tolerated
Prefer unencrypted peer connections.
.It Fl f, -finish Ar script
Set a script to run when the torrent finishes
.It Fl g, Fl -config-dir Ar directory
Where to look for configuration files.  This can be used to swap between using the cli, daemon, gtk, and qt clients.
See http://trac.transmissionbt.com/wiki/ConfigFiles for more information.
.It Fl h, Fl -help
Prints a short usage summary.
.It Fl i, Fl -info
Shows torrent details and exit
.It Fl m, Fl -portmap
Enable portmapping via NAT-PMP or UPnP
.It Fl M, Fl -no-portmap
Disable portmapping
.It Fl n, Fl -new Ar sourcefile
Create torrent from the specified file or directory
.It Fl p, -port Ar port
Set the port to listen for incoming peers. (Default: 51413)
.It Fl r, Fl -private
When creating a new torrent, set its 'private' flag
.It Fl s, -scrape
Print the current number of seeders and leechers for the specified torrent
.It Fl t, -tos
Set the peer socket TOS for local router-based traffic shaping.
.It Fl u, -uplimit Ar number
Set the maximum upload speed in KB/s
.It Fl U, -no-uplimit
Don't limit the upload speed
.It Fl v, Fl -verify
Verify the torrent's downloaded data.
.It Fl w, Fl -download-dir Ar directory
Where to save downloaded data.
.El
.Sh SIGNALS
In addition to these options, sending
.Nm
a SIGHUP signal will contact the tracker for more peers.
.Sh FILES
.Bl -tag -width Ds -compact
.It ~/.config/transmission
Directory where
.Nm
keeps torrent information for future seeding and resume operations.
.El
.Sh AUTHORS
The
.Nm
program was written by 
.An -nosplit
.An Eric Petit ,
.An Josh Elsasser ,
.An Charles Kerr ,
and
.An Mitchell Livingston .
.Sh SEE ALSO
.Xr transmission 1 ,
.Xr transmission-daemon 1 ,
.Xr transmission-remote 1
.Pp
http://www.transmissionbt.com/


NEWS
=== Transmission 2.13 (2010/12/09) ===
[http://trac.transmissionbt.com/query?milestone=2.13&group=component&order=severity All tickets closed by this release]
==== All Platforms ====
  * Fix "No Announce Scheduled" tracker error
  * Fix build error on Solaris 10
  * Fix RPC documentation errors
  * Fix minor packaging errors
==== Mac ====
  * Add option to run a script when torrents finish downloading
  * Simplify editing the blocklist URL in the preferences window
==== GTK+ ====
  * Unblur the statusbar icons
  * Fix truncation error in the Torrent Properties dialog.
==== Qt ====
  * Fix crash when opening the Torrent Properties dialog on magnet links
  * Fix "undo" error when making changes in the Torrent Properties dialog
  * Add Brazilian Portuguese Translation
  * Add Spanish (LAC) Translation

=== Transmission 2.12 (2010/11/14) ===
[http://trac.transmissionbt.com/query?milestone=2.12&group=component&order=severity All tickets closed by this release]
==== All Platforms ====
  * Make the blocklist URL user-configurable
  * Several minor bug fixes and CPU speed improvements
  * Use slightly less bandwidth in DHT messages
  * Fix 2.10 build issue on uClibc systems
==== Mac ====
  * Sort the file list alphabetically
  * Ensure the proper extension is used when saving the torrent file
  * Allow the Quick Look command when the inspector window has focus
==== GTK+ ====
  * Fix 2.11 crash when opening the Properties dialog on a magnet link torrent
  * Fix 2.00 regression which failed to inhibit hibernation on laptops
==== Qt ====
  * Fix 2.10 build issue on Ubuntu
==== CLI ====
  * Fix 2.10 crash
==== Web Client ====
  * Upgrade to jQuery 1.4.3
  * Fix 2.11 regression when using the web client on IE7 or IE8

=== Transmission 2.11 (2010/10/16) ===
[http://trac.transmissionbt.com/query?milestone=2.11&group=component&order=severity All tickets closed by this release]
==== All Platforms ====
  * Fix a 2.10 crash
  * Fix minor build issues on various platforms, particularly embedded systems
  * Fix issue when saving bencoded files on small hardware running uClibc
  * Fix minor rounding issue when displaying percentages
  * Fix the optimistic unchoke interval
==== Mac ====
  * Swipe to change inspector tab
==== GTK+ ====
  * Fix drag-and-dropping a magnet link
  * Fix hiding dialogs when hiding Transmission in the notification area
==== Qt ====
  * Fix crash in the file tree
==== Daemon ====
  * Fix missing status message when using "transmission-remote --add"
==== Web Client ====
  * Add Transmission website links to the action menu

=== Transmission 2.10 (2010/10/07) ===
[http://trac.transmissionbt.com/query?milestone=2.10&group=component&order=severity All tickets closed by this release]
==== All Platforms ====
  * Support for automatically stopping idle seeding torrents
  * Small memory cache to reduce disk IO
  * gzip compression support added to the RPC server
  * Add support for a cookies.txt file
  * Minor CPU optimizations
==== Mac ====
  * Display file sizes and speeds in base 10 on Snow Leopard
==== GTK+ ====
  * Files and folders can be opened by clicking on them in the files list
  * Update the interface when session changes are made via RPC
  * Fix the Details dialog to fit on a netbook screen
==== Qt ====
  * Tracker announce list editing
  * New filterbar
  * Improved display for showing a torrent's tracker announces
  * Better DBUS integration
  * Support adding torrents via drag-and-drop
  * Add Desktop Notification for added/complete torrents
  * Other minor improvements
==== Web Client ====
  * Peer list added to the inspector
  * Compact view mode
  * Support filtering by active and finished
  * Support sorting by size
==== Daemon ====
  * Allow the .pidfile location to be set in settings.json
==== Utils ====
  * New command-line utility "transmission-edit" for editing torrent files
  * New command-line utility "transmission-show" for viewing torrent files
  * New command-line utility "transmission-create" for creating torrent files

=== Transmission 2.04 (2010/08/06) ===
[http://trac.transmissionbt.com/query?milestone=2.04&group=component&order=severity All tickets closed by this release]
==== All Platforms ====
  * Improved shutdown time in some situations
  * Fix display bug in tr_truncd()
==== Mac ====
  * Fix bug that caused Local Peer Discovery to always be disabled on startup
==== GTK+ ====
  * Fix inaccurate "active torrent" counts in the filterbar
  * Fix display bug with magnet link names
==== Qt ====
  * Fix crash when accessing a password-protected remote session
==== Web Client ====
  * Fix bad redirect

=== Transmission 2.03 (2010/07/21) ===
[http://trac.transmissionbt.com/query?milestone=2.03&group=component&order=severity All tickets closed by this release]
==== Mac ====
  * Fix 2.02 bug where new transfers could not be added and the inspector would not appear

=== Transmission 2.02 (2010/07/19) ===
[http://trac.transmissionbt.com/query?milestone=2.02&group=component&order=severity All tickets closed by this release]
==== All Platforms ====
  * Fix NAT-PMP port forwarding issue on some systems
  * Fix filename errors for torrents whose files are encoded in ISO-8859-1
  * Fix rare crash on shutdown
  * Fix the RPC server's redirect URL to allow HTTPS proxies like stunnel
  * Replace less-portable calls with posix nanosleep()
==== Mac ====
  * Use F_NOCACHE to keep "inactive memory" in check
==== GTK+ ====
  * Fix crash when opening the Properties dialog on a magnet link without metainfo
  * Fix crash when removing multiple torrents at once
  * Allow individual torrents' download speed limits to be set to zero
  * Fix translation error with some error messages
==== Qt ====
  * Fix CPU spike when opening the Properties dialog

cli/transmissioncli.c
/******************************************************************************
 * $Id$
 *
 * Copyright (c) 2005-2006 Transmission authors and contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>

#include <libtransmission/transmission.h>
#include <libtransmission/bencode.h>
#include <libtransmission/makemeta.h>
#include <libtransmission/metainfo.h> /* tr_metainfoFree */
#include <libtransmission/tr-getopt.h>
#include <libtransmission/utils.h> /* tr_wait */
#include <libtransmission/web.h> /* tr_webRun */

#define MY_NAME "transmission-cli"

static int           showInfo         = 0;
static int           showScrape       = 0;
static int           isPrivate        = 0;
static int           verboseLevel     = 0;
static int           encryptionMode   = TR_ENCRYPTION_PREFERRED;
static int           peerPort         = TR_DEFAULT_PORT;
static int           peerSocketTOS    = TR_DEFAULT_PEER_SOCKET_TOS;
static int           blocklistEnabled = TR_DEFAULT_BLOCKLIST_ENABLED;
static int           uploadLimit      = 20;
static int           downloadLimit    = -1;
static int           natTraversal     = TR_DEFAULT_PORT_FORWARDING_ENABLED;
static int           verify           = 0;
static sig_atomic_t  gotsig           = 0;
static sig_atomic_t  manualUpdate     = 0;

static const char   * torrentPath   = NULL;
static const char   * downloadDir   = NULL;
static const char   * finishCall   = NULL;
static const char   * announce     = NULL;
static const char   * configdir    = NULL;
static const char   * sourceFile   = NULL;
static const char   * comment      = NULL;

static int  parseCommandLine ( int argc, const char ** argv );
static void sigHandler       ( int signal );

static char *
getStringRatio( float ratio )
{
    static char string[20];

    if( ratio == TR_RATIO_NA )
        return "n/a";
    snprintf( string, sizeof string, "%.3f", ratio );
    return string;
}

static int
is_rfc2396_alnum( char ch )
{
    return     ( '0' <= ch && ch <= '9' )
            || ( 'A' <= ch && ch <= 'Z' )
            || ( 'a' <= ch && ch <= 'z' );
}

static void
escape( char * out, const uint8_t * in, int in_len ) /* rfc2396 */
{
    const uint8_t *end = in + in_len;
    while( in != end )
        if( is_rfc2396_alnum(*in) )
            *out++ = (char) *in++;
        else
            out += snprintf( out, 4, "%%%02X", (unsigned int)*in++ );
    *out = '\0';
}


#define LINEWIDTH 80

static void
torrentStateChanged( tr_torrent   * torrent UNUSED,
                     cp_status_t    status UNUSED,
                     void         * user_data UNUSED )
{
    system( finishCall );
}

static int leftToScrape = 0;

static void
scrapeDoneFunc( struct tr_handle    * session UNUSED,
                long                  response_code,
                const void          * response,
                size_t                response_byte_count,
                void                * host )
{
    tr_benc top, *files;

    if( !tr_bencLoad( response, response_byte_count, &top, NULL ) 
        && tr_bencDictFindDict( &top, "files", &files )
        && files->val.l.count >= 2 )
    {
        int64_t complete=-1, incomplete=-1, downloaded=-1;
        tr_benc * hash = &files->val.l.vals[1];
        tr_bencDictFindInt( hash, "complete", &complete );
        tr_bencDictFindInt( hash, "incomplete", &incomplete );
        tr_bencDictFindInt( hash, "downloaded", &downloaded );
        printf( "%4d seeders, %4d leechers, %5d downloads at %s\n",
                (int)complete, (int)incomplete, (int)downloaded, (char*)host );
        tr_bencFree( &top );
    }
    else
        printf( "unable to parse response (http code %lu) at %s", response_code, (char*)host );

    --leftToScrape;
}

int
main( int argc, char ** argv )
{
    int i, error;
    tr_handle  * h;
    tr_ctor * ctor;
    tr_torrent * tor = NULL;
    char cwd[MAX_PATH_LENGTH];

    printf( "Transmission %s - http://www.transmissionbt.com/\n",
            LONG_VERSION_STRING );

    /* Get options */
    if( parseCommandLine( argc, (const char**)argv ) )
        return EXIT_FAILURE;

    /* Check the options for validity */
    if( !torrentPath ) {
        printf( "No torrent specified!\n" );
        return EXIT_FAILURE;
    }
    if( peerPort < 1 || peerPort > 65535 ) {
        printf( "Invalid port '%d'\n", peerPort );
        return EXIT_FAILURE;
    }
    if( peerSocketTOS < 0 || peerSocketTOS > 255 ) {
        printf( "Invalid TOS '%d'\n", peerSocketTOS );
        return EXIT_FAILURE;
    }

    /* don't bind the port if we're just running the CLI 
     * to get metainfo or to create a torrent */
    if( showInfo || showScrape || ( sourceFile != NULL ) )
        peerPort = -1;

    if( configdir == NULL )
        configdir = tr_getDefaultConfigDir( );

    /* if no download directory specified, use cwd instead */
    if( !downloadDir ) {
        getcwd( cwd, sizeof( cwd ) );
        downloadDir = cwd;
    }


    /* Initialize libtransmission */
    h = tr_sessionInitFull(
            configdir,
            "cli",                         /* tag */
            downloadDir,                   /* where to download torrents */
            TR_DEFAULT_PEX_ENABLED,
            natTraversal,                  /* nat enabled */
            peerPort,
            encryptionMode,
            uploadLimit >= 0,
            uploadLimit,
            downloadLimit >= 0,
            downloadLimit,
            TR_DEFAULT_GLOBAL_PEER_LIMIT,
            verboseLevel + 1,              /* messageLevel */
            0,                             /* is message queueing enabled? */
            blocklistEnabled,
            peerSocketTOS,
            TR_DEFAULT_RPC_ENABLED,
            TR_DEFAULT_RPC_PORT,
            TR_DEFAULT_RPC_ACL,
            FALSE, "fnord", "potzrebie",
            TR_DEFAULT_PROXY_ENABLED,
            TR_DEFAULT_PROXY,
            TR_DEFAULT_PROXY_TYPE,
            TR_DEFAULT_PROXY_AUTH_ENABLED,
            TR_DEFAULT_PROXY_USERNAME,
            TR_DEFAULT_PROXY_PASSWORD );

    if( sourceFile && *sourceFile ) /* creating a torrent */
    {
        int err;
        tr_metainfo_builder * builder = tr_metaInfoBuilderCreate( h, sourceFile );
        tr_tracker_info ti;
        ti.tier = 0;
        ti.announce = (char*) announce;
        tr_makeMetaInfo( builder, torrentPath, &ti, 1, comment, isPrivate );
        while( !builder->isDone ) {
            tr_wait( 1000 );
            printf( "." );
        }
        err = builder->result;
        tr_metaInfoBuilderFree( builder );
        return err;
    }

    ctor = tr_ctorNew( h );
    tr_ctorSetMetainfoFromFile( ctor, torrentPath );
    tr_ctorSetPaused( ctor, TR_FORCE, showScrape );
    tr_ctorSetDownloadDir( ctor, TR_FORCE, downloadDir );

    if( showScrape )
    {
        tr_info info;

        if( !tr_torrentParse( h, ctor, &info ) )
        {
            int i;
            const time_t start = time( NULL );
            for( i=0; i<info.trackerCount; ++i )
            {
                if( info.trackers[i].scrape )
                {
                    const char * scrape = info.trackers[i].scrape;
                    char escaped[SHA_DIGEST_LENGTH*3 + 1];
                    char *url, *host;
                    escape( escaped, info.hash, SHA_DIGEST_LENGTH );
                    url = tr_strdup_printf( "%s%cinfo_hash=%s",
                                            scrape,
                                            strchr(scrape,'?')?'&':'?',
                                            escaped );
                    tr_httpParseURL( scrape, -1, &host, NULL, NULL );
                    ++leftToScrape;
                    tr_webRun( h, url, NULL, scrapeDoneFunc, host );
                    tr_free( url );
                }
            }

            fprintf( stderr, "scraping %d trackers:\n", leftToScrape );

            while( leftToScrape>0 && ((time(NULL)-start)<20) )
                tr_wait( 250 );
        }
        goto cleanup;
    }

    if( showInfo )
    {
        tr_info info;

        if( !tr_torrentParse( h, ctor, &info ) )
        {
            int prevTier = -1;
            tr_file_index_t ff;

            printf( "hash:\t" );
            for( i=0; i<SHA_DIGEST_LENGTH; ++i )
                printf( "%02x", info.hash[i] );
            printf( "\n" );

            printf( "name:\t%s\n", info.name );

            for( i=0; i<info.trackerCount; ++i ) {
                if( prevTier != info.trackers[i].tier ) {
                    prevTier = info.trackers[i].tier;
                    printf( "\ntracker tier #%d:\n", (prevTier+1) );
                }
                printf( "\tannounce:\t%s\n", info.trackers[i].announce );
            }

            printf( "size:\t%"PRIu64" (%"PRIu64" * %d + %"PRIu64")\n",
                    info.totalSize, info.totalSize / info.pieceSize,
                    info.pieceSize, info.totalSize % info.pieceSize );

            if( info.comment[0] )
                printf( "comment:\t%s\n", info.comment );
            if( info.creator[0] )
                printf( "creator:\t%s\n", info.creator );
            if( info.isPrivate )
                printf( "private flag set\n" );

            printf( "file(s):\n" );
            for( ff=0; ff<info.fileCount; ++ff )
                printf( "\t%s (%"PRIu64")\n", info.files[ff].name, info.files[ff].length );

            tr_metainfoFree( &info );
        }

        tr_ctorFree( ctor );
        goto cleanup;
    }

    tor = tr_torrentNew( h, ctor, &error );
    tr_ctorFree( ctor );
    if( tor == NULL )
    {
        printf( "Failed opening torrent file `%s'\n", torrentPath );
        tr_sessionClose( h );
        return EXIT_FAILURE;
    }

    signal( SIGINT, sigHandler );
    signal( SIGHUP, sigHandler );

    tr_torrentSetStatusCallback( tor, torrentStateChanged, NULL );
    tr_torrentStart( tor );

    for( ;; )
    {
        char string[LINEWIDTH];
        int  chars = 0;
        const struct tr_stat * s;

        tr_wait( 1000 );

        if( gotsig )
        {
            gotsig = 0;
            tr_torrentStop( tor );
            tr_sessionSetPortForwardingEnabled( h, 0 );
        }
        
        if( manualUpdate )
        {
            manualUpdate = 0;
            if ( !tr_torrentCanManualUpdate( tor ) )
                fprintf( stderr, "\rReceived SIGHUP, but can't send a manual update now\n" );
            else {
                fprintf( stderr, "\rReceived SIGHUP: manual update scheduled\n" );
                tr_torrentManualUpdate( tor );
            }
        }
        
        if( verify )
        {
            verify = 0;
            tr_torrentVerify( tor );
        }

        s = tr_torrentStat( tor );

        if( s->status & TR_STATUS_CHECK_WAIT )
        {
            chars = snprintf( string, sizeof string,
                "Waiting to verify local files..." );
        }
        else if( s->status & TR_STATUS_CHECK )
        {
            chars = snprintf( string, sizeof string,
                "Verifying local files... %.2f%%, found %.2f%% valid", 100 * s->recheckProgress, 100.0 * s->percentDone );
        }
        else if( s->status & TR_STATUS_DOWNLOAD )
        {
            chars = snprintf( string, sizeof string,
                "Progress: %.2f %%, %d peer%s, dl from %d (%.2f KB/s), "
                "ul to %d (%.2f KB/s) [%s]", 100.0 * s->percentDone,
                s->peersConnected, ( s->peersConnected == 1 ) ? "" : "s",
                s->peersSendingToUs, s->rateDownload,
                s->peersGettingFromUs, s->rateUpload,
                getStringRatio(s->ratio) );
        }
        else if( s->status & TR_STATUS_SEED )
        {
            chars = snprintf( string, sizeof string,
                "Seeding, uploading to %d of %d peer(s), %.2f KB/s [%s]",
                s->peersGettingFromUs, s->peersConnected,
                s->rateUpload, getStringRatio(s->ratio) );
        }
        else if( s->status & TR_STATUS_STOPPED )
        {
            break;
        }
        if( ( signed )sizeof string > chars )
        {
            memset( &string[chars], ' ', sizeof string - 1 - chars );
        }
        string[sizeof string - 1] = '\0';
        fprintf( stderr, "\r%s", string );

        if( s->error )
        {
            fprintf( stderr, "\n%s\n", s->errorString );
        }
        else if( verboseLevel > 0 )
        {
            fprintf( stderr, "\n" );
        }
    }
    fprintf( stderr, "\n" );

    /* try for 5 seconds to delete any port mappings for nat traversal */
    tr_sessionSetPortForwardingEnabled( h, 0 );
    for( i=0; i<10; ++i ) {
        const tr_port_forwarding f = tr_sessionGetPortForwarding( h );
        if( f == TR_PORT_UNMAPPED )
            break;
        tr_wait( 500 );
    }
    
cleanup:
    tr_sessionClose( h );

    return EXIT_SUCCESS;
}

/***
****
****
****
***/

static const char *
getUsage( void )
{
    return "A fast and easy BitTorrent client\n"
           "\n"
           "Usage: "MY_NAME" [options] <torrent-filename>";
}

const struct tr_option options[] = {
    { 'a', "announce",     "When creating a new torrent, set its announce URL",    "a", 1, "<url>" },
    { 'b', "blocklist",    "Enable peer blocklists",                               "b", 0, NULL },
    { 'B', "no-blocklist", "Disable peer blocklists",                              "B", 0, NULL },
    { 'c', "comment",      "When creating a new torrent, set its comment field",   "c", 1, "<comment>" },
    { 'd', "downlimit",    "Set the maxiumum download speed in KB/s",              "d", 1, "<number>" },
    { 'D', "no-downlimit", "Don't limit the download speed",                       "D", 0, NULL },
    { 'e', "encryption",   "Set encryption mode [required, preferred, tolerated]", "e", 1, "<mode>" },
    { 'f', "finish",       "Set a script to run when the torrent finishes",        "f", 1, "<script>" },
    { 'g', "config-dir",   "Where to look for configuration files",                "g", 1, "<path>" },
    { 'i', "info",         "Show torrent details and exit",                        "i", 0, NULL },
    { 'm', "portmap",      "Enable portmapping via NAT-PMP or UPnP",               "m", 0, NULL },
    { 'M', "no-portmap",   "Disable portmapping",                                  "M", 0, NULL },
    { 'n', "new",          "Create a new torrent from a file or directory",        "n", 1, "<path>" },
    { 'p', "port",         "Port to listen for incoming peers (Default: "TR_DEFAULT_PORT_STR")", "p", 1, "<port>" },
    { 'r', "private",      "When creating a new torrent, set its 'private' flag",  "r", 0, NULL },
    { 's', "scrape",       "Scrape the torrent and exit",                          "s", 0, NULL },
    { 't', "tos",          "Peer socket TOS (0 to 255, default="TR_DEFAULT_PEER_SOCKET_TOS_STR")", "t", 1, "<number>"},
    { 'u', "uplimit",      "Set the maxiumum upload speed in KB/s",                "u", 1, "<number>" },
    { 'U', "no-uplimit",   "Don't limit the upload speed",                         "U", 0, NULL },
    { 'v', "verify",       "Verify the specified torrent",                         "v", 0, NULL },
    { 'w', "download-dir", "Where to save downloaded data",                        "w", 1, "<path>" },
    { 0, NULL, NULL, NULL, 0, NULL }
};

static void
showUsage( void )
{
    tr_getopt_usage( MY_NAME, getUsage(), options );
    exit( 0 );
}

static int
numarg( const char * arg )
{
    char * end = NULL;
    const long num = strtol( arg, &end, 10 );
    if( *end ) {
        fprintf( stderr, "Not a number: \"%s\"\n", arg );
        showUsage( );
    }
    return num;
}

static int
parseCommandLine( int argc, const char ** argv )
{
    int c;
    const char * optarg;

    while(( c = tr_getopt( getUsage(), argc, argv, options, &optarg )))
    {
        switch( c )
        {
            case 'a': announce = optarg; break;
            case 'b': blocklistEnabled = 1; break;
            case 'B': blocklistEnabled = 0; break;
            case 'c': comment = optarg; break;
            case 'd': downloadLimit = numarg( optarg ); break;
            case 'D': downloadLimit = -1; break;
            case 'e': if( !strcmp( optarg, "required" ) )
                          encryptionMode = TR_ENCRYPTION_REQUIRED;
                      else if( !strcmp( optarg, "tolerated" ) )
                          encryptionMode = TR_PLAINTEXT_PREFERRED;
                      else
                          encryptionMode = TR_ENCRYPTION_PREFERRED;
                      break;
            case 'f': finishCall = optarg; break;
            case 'g': configdir = optarg; break;
            case 'i': showInfo = 1; break;
            case 'm': natTraversal = 1; break;
            case 'M': natTraversal = 0; break;
            case 'n': sourceFile = optarg; break;
            case 'p': peerPort = numarg( optarg ); break;
            case 'r': isPrivate = 1; break;
            case 's': showScrape = 1; break;
            case 't': peerSocketTOS = numarg( optarg ); break;
            case 'u': uploadLimit = numarg( optarg ); break;
            case 'U': uploadLimit = -1; break;
            case 'v': verify = 1; break;
            case 'w': downloadDir = optarg; break;
            case TR_OPT_UNK: torrentPath = optarg; break;
            default: return 1;
        }
    }

    return 0;
}

static void
sigHandler( int signal )
{
    switch( signal )
    {
        case SIGINT: gotsig = 1; break;
        case SIGHUP: manualUpdate = 1; break;
        default: break;
    }
}
